### PR TITLE
implement ziti_context level events

### LINKED ISSUE #68: Would like a method or callback for status of the network controller
Is it online/available?  In an error state?  Is the status current or is it from some timeAgo?

### PR DESCRIPTION
**Pull request by ekoby**
[fixes #68]

- `ziti_event_cb` API - done
- `ZitiContextEvent`  - done
- `ZitiServiceEvent` - done
- `ZitiEdgeRouterEvent` - done

### CODE REVIEW

File: inc_internal/zt_internal.h [hunk 1]
```diff
@@ -47,6 +47,8 @@ typedef void (*send_cb)(int status, void *ctx);
 
 typedef void (*ch_connect_cb)(ziti_channel_t *ch, void *ctx, int status);
 
+typedef void (*ch_notify_state)(ziti_channel_t *ch, ziti_router_status status, void *ctx);
```
Comment (dovholuknf) [inline inc_internal/zt_internal.h:50]: maybe keep the parameter order the same? ish? keep the order the same as the ch_connect_cb ? (flip ctx/status)

Comment (ekoby) [reply inc_internal/zt_internal.h:50]: it is internal API, consistency is not required. Also I prefer `ctx` parameter to go last

File: includes/ziti/ziti_events.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+
+#ifndef ZITI_SDK_ZITI_EVENTS_H
```
Comment (dovholuknf) [inline includes/ziti/ziti_events.h:18]: this probably all needs some sweet doxygen comments

File: includes/ziti/ziti.h [hunk 3]
```diff
@@ -236,21 +239,27 @@ typedef struct ziti_options_s {
...
-    void *ctx;
-
     //posture query cbs
     ziti_pq_mac_cb pq_mac_cb;
     ziti_pq_os_cb pq_os_cb;
     ziti_pq_process_cb pq_process_cb;
     ziti_pq_domain_cb pq_domain_cb;
+
+    void *app_ctx;
+    unsigned int events;
```
Comment (smilindave26) [inline includes/ziti/ziti.h:261]: `events_mask` a better name? (or some doxygen)

File: includes/ziti/ziti.h [hunk 4]
```diff
@@ -236,21 +239,27 @@ typedef struct ziti_options_s {
...
 
-    void *ctx;
-
     //posture query cbs
     ziti_pq_mac_cb pq_mac_cb;
     ziti_pq_os_cb pq_os_cb;
     ziti_pq_process_cb pq_process_cb;
     ziti_pq_domain_cb pq_domain_cb;
+
+    void *app_ctx;
```
Comment (smilindave26) [inline includes/ziti/ziti.h:260]: would be nice to have a getter for this (similar to what you see in libuv)

Comment (dovholuknf) [inline includes/ziti/ziti.h:260]: shouldn't all the other callbacks get this context too? or could we somehow bake it into the ziti_context?

File: includes/ziti/ziti_events.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+typedef enum {
+    EdgeRouterConnected,
+    EdgeRouterDisconnected,
+    EdgeRouterRemoved,
+    EdgeRouterUnavailable,
+} ziti_router_status;
+
+struct ziti_context_event {
+    int ctrl_status;
```
Comment (smilindave26) [inline includes/ziti/ziti_events.h:39]: just an `int`? (no type like you see for e.g., `ziti_router_status`)

Comment (dovholuknf) [reply includes/ziti/ziti_events.h:39]: should it be renamed to `ziti_ctrl_event` for consistency then too? i like the idea of "router events" and "controller" events in that way. there's also "service events" so it's consistent

File: includes/ziti/ziti_events.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+struct ziti_router_event {
+    ziti_router_status status;
+    const char* name;
+    const char* version;
+};
+
+struct ziti_service_event {
+    ziti_service_array removed;
+    ziti_service_array changed;
+    ziti_service_array added;
```
Comment (dovholuknf) [inline includes/ziti/ziti_events.h:52]: should added events be here? or should those still come by `service_cb` from `init_opts`? seems like a double dip to me? or are events replacing the `service_cb` entirely? i'm not sure how these two are supposed to correlate

Comment (ekoby) [reply includes/ziti/ziti_events.h:87]: service_cb is deprecated, for now both are supported

File: includes/ziti/ziti_events.h [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * \brief Object passed to `ziti_options.event_cb`.
+ *
+ * \note event data becomes invalid as soon as callback returns.
+ * App must copy data if it's needed for further processing.
+ */
+typedef struct ziti_event_s {
+    ziti_event_type type;
+    union {
```
Comment (smilindave26) [inline includes/ziti/ziti_events.h:98]: Not sure how this will go with Swift (use of `union`)...

Comment (ekoby) [reply includes/ziti/ziti_events.h:98]: let me know :)

File: includes/ziti/ziti.h (hunk 3 above)
Comment (ekoby) [reply includes/ziti/ziti.h:276]: doxygen added

File: includes/ziti/ziti.h (hunk 4 above)
Comment (ekoby) [reply includes/ziti/ziti.h:271]: added `ziti_app_context()`

Comment (dovholuknf) [reply includes/ziti/ziti.h:271]: are you leaving the context in the event callback? i feel like it doesn't belong on `ziti_event_cb` with this

**Review by scareything — APPROVED** [review]

**Review by smilindave26 — APPROVED** [review]

**Pull request merged**